File: src/main/go/leetcode/lists/offerII/Offer44.go
package offerII

import "math"

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/16 16:58
 */

func largestValues(root *TreeNode) []int {

	if root == nil {
		return []int{}
	}

	q := make([]*TreeNode, 0)
	q = append(q, root)
	res := make([]int, 0)
	for len(q) > 0 {
		size := len(q)
		max := math.MinInt32
		for size > 0 {
			node := q[0]
			q = q[1:]
			if max < node.Val {
				max = node.Val
			}
			if node.Left != nil {
				q = append(q, node.Left)
			}
			if node.Right != nil {
				q = append(q, node.Right)
			}
			size--
		}
		res = append(res, max)
	}
	return res
}

File: java/src/leetcode/lists/offerII/Offer44.java
package leetcode.lists.offerII;

import common.TreeNode;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/7/25 10:57
 */
public class Offer44 {


    public List<Integer> largestValues(TreeNode root) {
        if (root == null) {
            return new ArrayList<>();
        }
        Queue<TreeNode> queue = new LinkedList<>();
        List<Integer> list = new ArrayList<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            int max = Integer.MIN_VALUE;
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode node = queue.poll();
                max = Math.max(max, node.val);
                if (node.left != null) {
                    queue.add(node.left);
                }
                if (node.right != null) {
                    queue.add(node.right);
                }
            }
            list.add(max);
            max = Integer.MIN_VALUE;
        }
        return list;
    }
}
